src/lib.rs

```diff
@@ -1 +1,452 @@
+    }
+}
+
+struct ResonanceState(Mutex<original::ResonanceState>);
+impl ResonanceState {
+    fn new(state: original::ResonanceState) -> Self {
+        ResonanceState(Mutex::new(state))
+    }
+}
+
+impl sgp4::ResonanceState for ResonanceState {
+    fn t(&self) -> f64 {
+        self.0.lock().expect("The mutex was poisioned").t()
+    }
+}
+
+struct Constants(original::Constants);
+
+impl Constants {
+    fn new_wrap(state: original::Constants) -> Self {
+        Constants(state)
+    }
+}
+
+impl sgp4::Constants for Constants {
+    fn new(
+        geopotential: Geopotential,
+        epoch_to_sidereal_time: EpochToSiderealTimeAlgorithm,
+        epoch: f64,
+        drag_item: f64,
+        orbit0: Orbit,
+    ) -> Result<Handle<Constants>, SgpError> {
+        let epoch_to_sidereal_time_fn = match epoch_to_sidereal_time {
+            EpochToSiderealTimeAlgorithm::Afspc => original::afspc_epoch_to_sidereal_time,
+            EpochToSiderealTimeAlgorithm::Iau => original::iau_epoch_to_sidereal_time,
+        };
+
+        match original::Constants::new(
+            geopotential.into(),
+            epoch_to_sidereal_time_fn,
+            epoch,
+            drag_item,
+            orbit0.into(),
+        ) {
+            Ok(state) => Ok(Handle::new(Constants::new_wrap(state))),
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    fn from_elements(elements: ElementS) -> Result<Handle<Constants>, SgpError> {
+        match original::Constants::from_elements(&elements.into()) {
+            Ok(state) => Ok(Handle::new(Constants::new_wrap(state))),
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    fn from_elements_afspc_compatibility_mode(
+        elements: ElementS,
+    ) -> Result<Handle<Constants>, SgpError> {
+        match original::Constants::from_elements_afspc_compatibility_mode(&elements.into()) {
+            Ok(state) => Ok(Handle::new(Constants::new_wrap(state))),
+            Err(err) => Err(err.into()),
+        }
+    }
+    fn propagate(&self, t: f64) -> Result<Prediction, SgpError> {
+        match self.0.propagate(t) {
+            Ok(prediction) => Ok(prediction.into()),
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    fn propagate_afspc_compatibility_mode(&self, t: f64) -> Result<Prediction, SgpError> {
+        match self.0.propagate_afspc_compatibility_mode(t) {
+            Ok(prediction) => Ok(prediction.into()),
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    fn initial_state(&self) -> Option<Handle<ResonanceState>> {
+        match self.0.initial_state() {
+            Some(rs) => Some(Handle::new(ResonanceState::new(rs))),
+            None => None,
+        }
+    }
+
+    fn propagate_from_state(
+        &self,
+        t: f64,
+        state: Option<Handle<ResonanceState>>,
+        afspc_compatibility_mode: bool,
+    ) -> Result<Prediction, SgpError> {
+        let mut rs = *state
+            .expect("Unable to get resonance state")
```

Comment:
We should try to handle this gracefully instead of blowing up when `ResonanceState` isn't provided. 

Maybe you can do something with `.map()` so that when `state` is `None` we pass `None` into `propagate_from_state()`, and when it is `Some` we pass `Some(&mut original::ResonanceState)` in?